### ISSUE TITLE
Bounding box orientation incorrect on custom nuscenes formatted dataset

### ISSUE THREAD

**Issue by pkamat95** [open-mmlab/mmdetection3d#976]
Hi,

We've been using mmdetection3d to train a custom dataset that we've created that is in the nuscenes format, using our own classes and data. We can render the ground truth annotations using the nuscenes devkit, with [nuscenes.render_sample_data()](https://github.com/nutonomy/nuscenes-devkit/blob/c44366daea8bba29673943c1fc86d0bfbfb7a99e/python-sdk/nuscenes/nuscenes.py#L548), and this visualisation is correct.

![trainval__perception_camera__rgb_2_nuscenes_gt](https://user-images.githubusercontent.com/29537979/135127422-56d444d5-7400-4027-afc7-98093dcb5c1c.png)

We [altered nuscenes_converter.py](https://github.com/open-mmlab/mmdetection3d/blob/master/tools/data_converter/nuscenes_converter.py) to work with our dataset. As there were several instances of things being hard coded to work specifically with the nuscenes dataset, we had to change several things to get the converter working with our dataset.

We can use an altered version of [browse_datasets.py](https://github.com/open-mmlab/mmdetection3d/blob/master/tools/misc/browse_dataset.py) to visualise the ground truth annotations after the conversion. When we visualise these, the translation is correct, but the orientation of the box is incorrect. Since translation appears perfectly, this leads me to think that there is no issue with the ego or sensor rotation and that it's a problem with only the annotation orientation. We can also train on the dataset and get a high mAP, but the predictions, as expected, have the same orientation as the ground truth data when visualised using [mono_det_demo.py](https://github.com/open-mmlab/mmdetection3d/blob/master/demo/mono_det_demo.py)

![trainval__perception_camera__rgb_2_mmdet_gt](https://user-images.githubusercontent.com/29537979/135127457-47071290-7b72-4576-9e65-bdd269764bfa.png)

After spending a lot of time attempting to understand where this rotation issue is coming from, we're at a loss. Hopefully someone could offer some help with this issue. @ZwwWayne I see that you've done [this PR](https://github.com/open-mmlab/mmdetection3d/pull/744) which seems quite relevant to what we've been working on. Perhaps you have an idea of what is going on?

I have packaged a dataset with a few images, and the conversion and visualisation scripts we've been using.
[mmdetection3d.zip](https://github.com/open-mmlab/mmdetection3d/files/7245930/mmdetection3d.zip)

If someone gets the time to help us out with this, it would be very appreciated.

If you want to run the conversion and visualisation scripts, build and run the docker container using the latest repo, and use the data folder in the attached zip as the `DATA_DIR` in `docker run --gpus all --shm-size=8g -it -v {DATA_DIR}:/mmdetection3d/data mmdetection3d`. Then place the scoville folder from the attached zip in `/mmdetection/` in the docker instance. Dataset conversion and visualisation can then be done using the following commands:
- `python scoville/pencil_nuscenes_converter.py data/nuscenes_pencil_example/ data/nuscenes_pencil_example/`
- `python scoville/browse_dataset_pencil.py scoville/pencil_finetune.py --output-dir visualisations --task mono-det`

**Comment by bml1g12** [comment]
Aha, I didn't realise this (I assumed incorrectly aligned boxes would produce poor mAP, similar to the 2D case with a low IoU threshold). 

```
img_bbox_NuScenes/pencil_AP_dist_0.5: 0.8536, img_bbox_NuScenes/pencil_AP_dist_1.0: 0.8536, img_bbox_NuScenes/pencil_AP_dist_2.0: 0.8536, img_bbox_NuScenes/pencil_AP_dist_4.0: 0.8536, img_bbox_NuScenes/pencil_trans_err: 0.1198, img_bbox_NuScenes/pencil_scale_err: 0.9782, img_bbox_NuScenes/pencil_orient_err: 2.0613, img_bbox_NuScenes/pencil_vel_err: 1.0000, img_bbox_NuScenes/pencil_attr_err: 1.0000, img_bbox_NuScenes/mATE: 0.1198, img_bbox_NuScenes/mASE: 0.9782, img_bbox_NuScenes/mAOE: 2.0613, img_bbox_NuScenes/mAVE: 1.0000, img_bbox_NuScenes/mAAE: 1.0000, img_bbox_NuScenes/NDS: 0.5170, img_bbox_NuScenes/mAP: 0.8536
```
I am not too familiar with this metric: **img_bbox_NuScenes/mAOE: 2.0613** but if I understand correctly it would suggest that the orientation is on average about 2 radians (118 degrees) out?

If so that helps us a lot, it would guess that it suggests that the visualization is not the issue after all, but in fact there is something wrong the with the orientation of ground truth data being fed in mmdet library such that the model is learning to provide inappropriate orientation. 

The mystery still remains though as to why we can visualise the raw nuscenes ground data fine with the NuScenes visualiser. It would suggest there is a bug somewhere in our [nuscenes converter](https://github.com/open-mmlab/mmdetection3d/blob/master/tools/data_converter/nuscenes_converter.py) I think - can you spot anything that looks suspicious?

**Comment by bml1g12** [comment]
Tried rewriting our converter using the latest version as a starting point, but ended with the same issue :( Sadly we might have to abandon this unless anyone has any further ideas on how we could debug.

**Comment by VeeranjaneyuluToka** [comment]
Is there any solution for this? I am facing exactly same issue. When my yaw angle is zero, it works fine but if there is yaw angle, the visualization goes for toss as showed above. Would be great help if there is some solution for this. Thanks!

**Comment by VeeranjaneyuluToka** [comment]
FYI, I fixed this issue. I have taken yaw = -yaw While converting from json files to .pkl  file, when i took yaw without any changes, the GTs are getting drawn as expected.  So this is basically a change in my converting script, but predictions are in wrong orientation while running inference as i have raised the same in https://github.com/open-mmlab/mmdetection3d/issues/1287.

**Comment by VeeranjaneyuluToka** [comment]
In my case, predictions are wrong becuase my training set samples are wrong.
